### ISSUE TITLE
Query execution ends, but psycopg hangs

### ISSUE THREAD

**Issue by wstrausser** [psycopg/psycopg#777]
I am currently working on a schema migration script for our database. The script is primarily extracting data from several fields within a JSONB column and creating new columns from these values.

When I execute the script from psycopg, the call to `cursor.execute()` hangs for about 2 hours before throwing the following error:

```
  File "/home/vscode/.local/lib/python3.12/site-packages/psycopg/cursor.py", line 732, in execute
    raise ex.with_traceback(None)
psycopg.OperationalError: consuming input failed: server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
```

After looking at `pg_stat_activity` during execution, I can see activity there for the first 5 or so minutes until the query disappears, meaning that psycopg is essentially sitting idle for just under 2 hours while appearing to be executing something.

It's possible that this is an issue with the database itself, but I've tested this exact same migration script from a separate environment (JetBrains DataGrip) and the full script executes without issue in about 20 minutes. As such, psycopg is all that I've been able to isolate the issue to thus far.

Worth also mentioning that I'm able to execute this script from psycopg on a local database with a small amount of data (<50k rows), but am hitting this issue on a development database with production-sized data (~10m rows).

**Comment by wstrausser** [comment]
Well, of course this issue has magically resolved itself and not sure why. I'll assume this is a database issue and come back here if it happens again. For now, I'll go ahead and close.

**Comment by wstrausser** [comment]
@dvarrazzo This issue actually popped again when I tried to run the same migration on our production database and I've now determined the cause.

What was happening was that the query would successfully execute in the database, but would take about 8 minutes to complete. By the time execution was completed, Postgres would try to notify psycopg, but encountered a broken pipe error, triggering postgres to close the connection. For whatever reason, psycopg was not picking up on the broken pipe and would continue to wait for a message from the database that never came.

The cause of the broken pipe error is related to the host's firewall configuration, which was closing idle connections after 5 minutes. I added a 60 second keepalive to my connection string and this completely resolved the issue.

Not sure if there's really anything that can be done on psycopg's side to address an issue like this since I don't understand the backend well enough and if there would be a way to pick up on the server closing a connection in this way. Hopefully this helps to clarify the issue, though.

Your point about JDBC is valid I think. If the keepalive didn't work, my next solution was going to be to try a server-side cursor, which I'm assuming would also have resolved this.
